lib/PromoteStackPointer/ComputeStackAccessesBoundsPass.cpp

```diff
@@ -2,58 +2,155 @@
+  LazyValueInfo &LVI = getAnalysis<LazyValueInfoWrapperPass>().getLVI();
+
+  if (F.isDeclaration())
+    return false;
+
   for (Function &StackOffsetFunction :
-       FunctionTags::StackOffsetMarker.functions(&M)) {
+       FunctionTags::StackOffsetMarker.functions(F.getParent())) {
+
     auto *FunctionType = StackOffsetFunction.getFunctionType();
     auto *DifferenceType = cast<IntegerType>(FunctionType->getParamType(1));
-    for (CallBase *Call : callers(&StackOffsetFunction)) {
-      Function &F = *Call->getParent()->getParent();
-      LazyValueInfo &LVI = getAnalysis<LazyValueInfoWrapperPass>(F).getLVI();
+
+    for (CallBase *Call : callersIn(&StackOffsetFunction, &F)) {
       auto *Undef = UndefValue::get(DifferenceType);
 
       // Identify lower bound of the lower bound
-      const auto &LowerBoundRange = LVI.getConstantRange(Call->getArgOperand(1),
-                                                         cast<CallInst>(Call));
-      Value *LowerBound = nullptr;
-      if (not LowerBoundRange.isFullSet()) {
-        LowerBound = ConstantInt::get(DifferenceType,
-                                      LowerBoundRange.getLower());
-      } else {
-        LowerBound = Undef;
-      }
-      Call->setArgOperand(1, LowerBound);
-
-      // Identify upper bound of the upper bound
-      const auto &UpperBoundRange = LVI.getConstantRange(Call->getArgOperand(2),
-                                                         cast<CallInst>(Call));
-      Value *UpperBound = nullptr;
-      if (not UpperBoundRange.isFullSet()) {
-        UpperBound = ConstantInt::get(DifferenceType,
-                                      UpperBoundRange.getUpper());
-      } else {
-        UpperBound = Undef;
-      }
-      Call->setArgOperand(2, UpperBound);
+      auto ReplaceUseWithBound = [&](Use &Operand, bool Lower) {
+        auto *User = Operand.getUser();
+        auto *V = Operand.get();
+        const APInt *Bound = nullptr;
+
+        // Get range from LVI
+        // TODO: using LVI is inaccurate (unlike using SCEV, the way we use it)
+        const auto &BoundRange = LVI.getConstantRange(V, cast<CallInst>(User));
+        if (not BoundRange.isFullSet()) {
+          Bound = Lower ? &BoundRange.getLower() : &BoundRange.getUpper();
+        }
+
+        if (auto *I = dyn_cast<Instruction>(V)) {
+          // See what SCEV says
+          auto SCEVRange = getSCEVBoundaries(SE, PDT, I);
+
+          APInt SCEVBound;
+          if (Lower) {
+            SCEVBound = SCEVRange.getLower();
+            if (not SCEVBound.isMinSignedValue()
+                and (Bound == nullptr or SCEVBound.sgt(*Bound))) {
+              Bound = &SCEVBound;
+            }
+          } else {
+            SCEVBound = SCEVRange.getUpper();
+            if (not SCEVBound.isMaxSignedValue()
+                and (Bound == nullptr or SCEVBound.sle(*Bound))) {
+              Bound = &SCEVBound;
+            }
+          }
+        }
```

Comment:
Also, you might want to use `llvm::APIntOps::smax(Bound, SCEVBound)` and `smin` instead of doing the weird dance with `sle`/`slt` and `sgt`. I think using them the `APIntOps` would have prevented the error in the first place.

Reply:
> AVI min

Note that AVI is not involved at all here. You probably meant LVI.

> At first it seems an error

Why? I've two candidate lower bounds and I'm looking for the strictest among the two. The strictest is the highest one. Viceversa for the upper bound.

> Also, it seems weird that you use sgt for Lower and sle (instead of slt) for the other case

I'll fix it for coherency purposes, but it's irrelevant. The case where both bounds are equal can both lead to keep the old value or update to the new one and makes no difference: they are the same.

> llvm::APIntOps::smax

Didn't know about that, I can use it but I just like the idea of having a code path that is taken only if we successfully found a stricter bound due to SCEV. For instance I can breakpoint there, add some logging.
But no strong opinions.